modules/javafx.base/src/main/java/javafx/beans/Observable.java

```diff
@@ -93,4 +94,22 @@ public interface Observable {
      */
     void removeListener(InvalidationListener listener);
 
+    /**
+     * Creates a {@link Subscription} on this value which calls the given
+     * {@code runnable} whenever it becomes invalid.
+     *
+     * @param subscriber a {@code Runnable} to call whenever this
+     *     value becomes invalid, cannot be {@code null}
+     * @return a {@code Subscription} which can be used to cancel this
+     *     subscription, never {@code null}
+     * @since 21
+     */
+    default Subscription invalidations(Runnable subscriber) {
```

Comment:
I think re-using `InvalidationListener` may fit better into the existing JavaFX API:
```java
interface Observable {
    Subscription subscribe(InvalidationListener);
}
```

The method can then be overloaded for `ChangeListener`, `ListChangeListener`, `SetChangeListener`, and `MapChangeListener` as well. Doing so allows users to use the same functional interfaces they've used before with the `addListener`/`removeListener` APIs.

As an added bonus, we don't need to allocate a wrapper for the listeners.

Reply:
This would depend I think how you want to approach this.  I added three variants, invalidations, changes and values.  Using the same name `subscribe`, you need to make sure the lambda's used all have a different arity, or you'll need to add a cast to disambiguate them.

I purposely removed the first `Observable` parameter on all of the variants because I think it will be a much better match when you want to register a listener, and "pipe" it to some existing code.  Now we often have to "strip" the unnecessary parameter.  I can't do this easily for example:

      minimumWidthProperty().addListener(this::requestLayout);
      selectedItem.addListener(getSelectionModel()::select);
      modelStatus.addListener(statusLabel.textProperty()::set);

It must be these now:

      minimumWidthProperty().addListener(obs -> this.requestLayout());
      selectedItem.addListener((obs, o, n) -> gelectionModel().select(n));
      modelStatus.addListener((obs, o, n) -> statusLabel.setText(n));

The `values` variant is highly useful for this, as `ChangeListener` has a 2nd rarely used parameter (oldValue).

Also I think calling unsubscribe on subscriptions should preferably be the only way to remove a listener again; reusing the existing listeners would allow you to bypass this with `removeListener`.

Reply:
These example would probably look like this in the near future:
``` java
minimumWidthProperty().addListener(this::requestLayout());
minimumWidthProperty().addListener(_ -> this.requestLayout());

selectedItem.addListener(getSelectionModel()::select);
selectedItem.addListener((_, _, value) -> gelectionModel().select(value));

modelStatus.addListener(statusLabel.textProperty()::set);
modelStatus.addListener((_, _, value) -> statusLabel.setText(value));
```

You can always ignore the `Observable` parameter, but you can't bring it back once it's gone. I have often used this parameter, for example when my listener handles changes of multiple similar properties. Removing functionality just to make it look a little prettier is not enough justification in my opinion.

If a subscription shouldn't be removable by passing the listener to `removeListener`, this can easily be achieved by internally wrapping it into another listener (just what you did, only with the functional interface being different). However, this might not be the best solution, as it doubles the number of listener objects when using the subscription API.

Reply:
Yeah I'm aware they want to add this, but it doesn't really invalidate my point as that syntax is terrible still.  Method references are far preferred whenever possible.

I stand by my point that using the observable parameter is a rare and quite advanced use case, where these API's would be intended to make it easier and safer to use method references and lambda's.

When dealing with multiple listeners that require the `Observable`, you are likely also dealing with adding and removing that single listener in some dynamic fashion.  Subscriptions don't work well with this as you'd need to track these, unlike `removeListener` which you can just pass the single listener reference.

Here's one of the only examples in JavaFX that uses the observable parameter (if you know of others, I'd be interested to see those as well):

        private final ChangeListener<Boolean> paneFocusListener = new ChangeListener<>() {
            @Override public void changed(ObservableValue<? extends Boolean> observable, Boolean oldValue, Boolean newValue) {
                if (newValue) {
                    final ReadOnlyBooleanProperty focusedProperty = (ReadOnlyBooleanProperty) observable;
                    final TitledPane tp = (TitledPane) focusedProperty.getBean();
                    focus(accordion.getPanes().indexOf(tp));
                }
            }
        };

And its management code:

        private final ListChangeListener<TitledPane> panesListener = c -> {
            while (c.next()) {
                if (c.wasAdded()) {
                    for (final TitledPane tp: c.getAddedSubList()) {
                        tp.focusedProperty().addListener(paneFocusListener);
                    }
                } else if (c.wasRemoved()) {
                    for (final TitledPane tp: c.getRemoved()) {
                        tp.focusedProperty().removeListener(paneFocusListener);
                    }
                }
            }
        };

To do this with `Subscription::unsubscribe` you'd need to track the `Subscriptions`... something like:

        private final Map<Property<?>, Subscription> subscriptions = new HashMap<>();
        private final ListChangeListener<TitledPane> panesListener = c -> {
            while (c.next()) {
                if (c.wasAdded()) {
                    for (final TitledPane tp: c.getAddedSubList()) {
                        subscriptions.put(tp.focusedProperty(), tp.focusedProperty().changes(paneFocusListener));
                    }
                } else if (c.wasRemoved()) {
                    for (final TitledPane tp: c.getRemoved()) {
                        subscriptions.remove(tp.focusedProperty()).unsubscribe();
                    }
                }
            }
        };

That more than likely goes against the reason for using a single listener in the first place.  Even if `removeListener` could be used still (a question that would need to be answered as it could limit our options later on), you'd be creating a `Subscription` only to discard it.  Now, I can imagine there might be use cases that don't need this kind of dynamic listener management but still need the observable parameter, but that would only make the use of that parameter in combination with subscriptions even more rare, and you'd still need the current API for some cases.

This is why I would really advocate for a simpler API, focusing on the use cases most often encountered by users of properties and controls, instead of cluttering these to also try to cater for control and property authors that (when dealing with performance optimizations that require the observable) are likely to need the existing methods anyway.

Reply:
I think the three newly added methods are a good choice. I wonder if we can some up with better names, though. without some verb like "add" or "subscribe" in the name, the name doesn't really indicate that it is adding a new listener to the observable.

Reply:
I agree that the chosen names `invalidation`, `changes` and `values` are a bit terse.  The whole signature (without reading docs) should make it clear you are creating a subscription, but perhaps we can do better.  The use of `addListener` can be ruled out as it would conflict with the existing method due to having Lambda's with the same arity (the `values` listener would conflict with `addListener(InvalidationListener)`.  Also, an `add` method would probably have users expecting a corresponding `remove` method.

A few ideas listed here:

| invalidation | values | changes |
|---|---|---|
|`subscribe(Runnable)`(*)|`subscribe(Consumer)`(*)|`subscribe(BiConsumer)`(*)|
|`subscribeInvalidations(Runnable)`|`subscribeValues(Consumer)`|`subscribeChanges(BiConsumer)`|
|`invalidationsTo(Runnable)`|`valuesTo(Consumer)`|`changesTo(BiConsumer)`|
(*) May limit future listener types that have same arity, but can still be a good choice

Reply:
On that same topic of naming methods:

What do people think of `Subscription#unsubscribe`?  Should it be `cancel`? Something else?  Okay as it is?

Code example:

        if (subscription != null) {
            subscription.unsubscribe();
            subscription = null;
        }

Reply:
> | invalidation | values | changes |
> |---|---|---|
> |`subscribe(Runnable)`(*)|`subscribe(Consumer)`(*)|`subscribe(BiConsumer)`(*)|
> |`subscribeInvalidations(Runnable)`|`subscribeValues(Consumer)`|`subscribeChanges(BiConsumer)`|
> |`invalidationsTo(Runnable)`|`valuesTo(Consumer)`|`changesTo(BiConsumer)`|
> 
> (*) May limit future listener types that have same arity, but can still be a good choice

My preference is in the order you listed them. If we go with `subscribe`, using method name overloading, do want to add a future overload that takes a type with the same arity as one of the existing three, we can always assign a new name to that new method (since adding another overload wouldn't be source compatible, we would likely need a new name at that point).

@nlisker @andy-goryachev-oracle @mstr2 - what do you think?

Reply:
> On that same topic of naming methods:
> 
> What do people think of `Subscription#unsubscribe`? Should it be `cancel`? Something else? Okay as it is?

OK as is. Especially if we go with one of the two choices with "subscribe" in the name.

Reply:
I might prefer the second option, subscribeXXX().
And unsubscribe() is good.

Reply:
I prefer `subscribe` with overloads, as this comes closest to the existing naming scheme of JavaFX listener operations (`addListener` / `removeListener` with overloads).

I would also like to see a subscription-style API that works with the existing listener types. This has the advantage of not needing a wrapper for each listener registration, and it can be a drop-in replacement in many places where JavaFX listeners are used today.

My suggestion is to have the following new operations:
* `subscribe` (`Runnable`, `Consumer`, `BiConsumer`)
* `subscribeListener` (`InvalidationListener`, `ChangeListener`)

This pairs nicely with the existing operations:
* `addListener` (`InvalidationListener`, `ChangeListener`)
* `removeListener` (`InvalidationListener`, `ChangeListener`)

Reply:
> My suggestion is to have the following new operations:
> 
> * `subscribe` (`Runnable`, `Consumer`, `BiConsumer`)

This seems reasonable. I could be talked into the longer names, but I prefer the overloaded `subscribe` name.

> * `subscribeListener` (`InvalidationListener`, `ChangeListener`)

This might be a useful follow-on Enhancement, but I wouldn't do it as part of this one.

Reply:
I've updated this PR now to use the `subscribe` naming, and created a corresponding CSR.